File: nightlife_engine/formatters/instagram_post_formatter.py
"""Instagram Post content formatter."""

from typing import Dict, Any
from .base_formatter import BaseFormatter


class InstagramPostFormatter(BaseFormatter):
    """
    Formats content for Instagram Posts.

    Output format:
    • Caption (max 2 sentences)
    • CTA line
    • Hashtags (exactly 5)
    """

    def format(self, generated_content: Dict[str, Any]) -> str:
        """Format Instagram Post content."""
        caption = generated_content.get("caption", "")
        cta = generated_content.get("cta", "")
        hashtags = generated_content.get("hashtags", [])

        # Ensure exactly 5 hashtags
        if len(hashtags) != 5:
            raise ValueError(f"Must have exactly 5 hashtags (got {len(hashtags)})")

        # Clean all parts
        caption = self._clean_output(caption)
        cta = self._clean_output(cta)
        hashtags = [self._clean_output(tag) for tag in hashtags]

        # Validate caption is max 2 sentences
        sentence_count = caption.count('.') + caption.count('!') + caption.count('?')
        if sentence_count > 2:
            raise ValueError(f"Caption must be max 2 sentences (got {sentence_count})")

        # Format output
        output_parts = [
            caption,
            "",
            cta,
            "",
            " ".join(hashtags)
        ]

        return "\n".join(output_parts)

File: webapp/lib/formatters/instagram-post-formatter.ts
/**
 * Instagram Post content formatter
 */

import { BaseFormatter, GeneratedContent } from "./base-formatter";

export class InstagramPostFormatter extends BaseFormatter {
  format(generatedContent: GeneratedContent): string {
    const caption = generatedContent.caption as string || "";
    const cta = generatedContent.cta as string || "";
    const hashtags = generatedContent.hashtags as string[] || [];

    // Ensure exactly 5 hashtags
    if (hashtags.length !== 5) {
      throw new Error(`Must have exactly 5 hashtags (got ${hashtags.length})`);
    }

    // Clean all parts
    const cleanedCaption = this.cleanOutput(caption);
    const cleanedCta = this.cleanOutput(cta);
    const cleanedHashtags = hashtags.map((tag) => this.cleanOutput(tag));

    // Validate caption is max 2 sentences
    const sentenceCount =
      (cleanedCaption.match(/[.!?]/g) || []).length;
    if (sentenceCount > 2) {
      throw new Error(
        `Caption must be max 2 sentences (got ${sentenceCount})`
      );
    }

    // Format output
    return [
      cleanedCaption,
      "",
      cleanedCta,
      "",
      cleanedHashtags.join(" "),
    ].join("\n");
  }
}
